### PR TITLE
raftstore-v2: fix "failed to get merge entries" panic

### LINKED ISSUE #15633: [Dynamic Regions] failed to get merge entires
### What version of TiKV are you using?

https://github.com/tikv/tikv/pull/15614   (Just added some debug logs)
commit: 267a9a00bfd3e1682fb2e4b88459f50dedcbcfd1

`
[FATAL] [lib.rs:510] ["failed to get merge entires [commit=130] [low=130] [err=Store(Compacted)] [peer_id=5817] [region_id=2581]"] [backtrace="   0: tikv_util::set_panic_hook::{{closu
re}}\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/lib.rs:509:18\n   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call\n             at
/rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2021:9\n      std::panicking::rust_panic_with_hook\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unk
nown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:733:13\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/r
ust/library/std/src/panicking.rs:621:13\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_co
mmon/backtrace.rs:170:18\n   4: rust_begin_unwind\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:617:5\n   5: core::panicking::panic_fmt\n
         at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:67:14\n   6: raftstore_v2::operation::command::admin::merge::commit::<impl raftstore_v2::raft::peer
::Peer<EK,ER>>::ask_target_peer_to_commit_merge\n      raftstore_v2::operation::command::admin::merge::commit::<impl raftstore_v2::raft::peer::Peer<EK,ER>>::on_check_merge\n             at /home/jenkins/agent/workspace/buil
d-common/go/src/github.com/pingcap/tikv/components/raftstore-v2/src/operation/command/admin/merge/commit.rs:167:9\n   7: raftstore_v2::fsm::peer::PeerFsmDelegate<EK,ER,T>::on_msgs\n             at /home/jenkins/agent/worksp
ace/build-common/go/src/github.com/pingcap/tikv/components/raftstore-v2/src/fsm/peer.rs:278:40\n   8: <raftstore_v2::batch::store::StorePoller<EK,ER,T> as batch_system::batch::PollHandler<raftstore_v2::fsm::peer::PeerFsm<EK
,ER>,raftstore_v2::fsm::store::StoreFsm>>::handle_normal\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore-v2/src/batch/store.rs:283:9\n      batch_system::batch
::Poller<N,C,Handler>::poll\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/batch-system/src/batch.rs:422:27\n   9: raftstore_v2::worker::refresh_config::PoolController<
N,C,H>::increase_by::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/raftstore-v2/src/worker/refresh_config.rs:75:21\n      <std::thread::Builder as tikv_uti
l::sys::thread::StdThreadBuildWrapper>::spawn_wrapper::{{closure}}\n             at /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tikv/components/tikv_util/src/sys/thread.rs:438:13\n      std::sys_com
mon::backtrace::__rust_begin_short_backtrace\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:154:18\n  10: std::thread::Builder::spa
wn_unchecked_::{{closure}}::{{closure}}\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:529:17\n      <core::panic::unwind_safe::AssertUnwindS
afe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:271:9\n      std::panicking::try
::do_call\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:524:40\n      std::panicking::try\n             at /rust/toolchains/nightly-2023-08-1
5-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:488:19\n      std::panic::catch_unwind\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/
std/src/panic.rs:142:14\n      std::thread::Builder::spawn_unchecked_::{{closure}}\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:528:30\n
   core::ops::function::FnOnce::call_once{{vtable.shim}}\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5\n  11: <alloc::boxed::Box<F,
A> as core::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2007:9\n      <alloc::boxed::Box<F,A> as core
::ops::function::FnOnce<Args>>::call_once\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2007:9\n      std::sys::unix::thread::Thread::new::thre
ad_start\n             at /rust/toolchains/nightly-2023-08-15-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17\n  12: start_thread\n  13: clone\n"] [location=/home/jenkins/agent/worksp
ace/build-common/go/src/github.com/pingcap/tikv/components/raftstore-v2/src/operation/command/admin/merge/commit.rs:199] [thread_name=rs-4-0]
`

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #15633

What's Changed:

```commit-message
fix "failed to get merge entries" panic
```
This PR fixes the panic of "failed to get merge entries".

Restart can lose merge status which in turns loses the limitations after apply PrepareMerge.  So, after start and becomes leader, it can propose CompactLog which may delete entries after `min_matched` which is required by CommitMerge.

This PR also fixes that region state will not be updated from `Merging` to `Normal` after apply Rollback Merge.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore-v2/src/fsm/peer.rs [hunk 1]
```diff
@@ -196,6 +196,7 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
         self.schedule_tick(PeerTick::SplitRegionCheck);
         self.schedule_tick(PeerTick::PdHeartbeat);
         self.schedule_tick(PeerTick::CompactLog);
+        self.fsm.peer.on_check_merge(self.store_ctx);
```
Comment (overvenus) [inline components/raftstore-v2/src/fsm/peer.rs:199]: How about adding `CheckMerge` tick? (to be consistent with others)
```suggestion
        self.schedule_tick(PeerTick::CheckMerge);
```

File: components/raftstore-v2/src/raft/peer.rs [hunk 2]
```diff
@@ -231,6 +231,11 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             unsafe_recovery_state: None,
         };
 
+        if let Some(ref state) = peer.merge_context {
```
Comment (overvenus) [inline components/raftstore-v2/src/raft/peer.rs:234]: Could you add some comments?

File: tests/failpoints/cases/test_merge.rs [hunk 3]
```diff
@@ -1853,3 +1843,126 @@ fn test_concurrent_between_transfer_leader_and_merge() {
...
+    configure_for_merge(&mut cluster.cfg);
+    cluster.cfg.raft_store.raft_log_gc_count_limit = Some(12);
+    cluster.cfg.raft_store.raft_log_gc_tick_interval = ReadableDuration::millis(10);
+    cluster.cfg.raft_store.raft_base_tick_interval = ReadableDuration::millis(10);
+    cluster.cfg.raft_store.merge_check_tick_interval = ReadableDuration::millis(10);
+
+    cluster.run();
+    fail::cfg("maybe_propose_compact_log", "return").unwrap();
+    fail::cfg("on_ask_commit_merge", "return").unwrap();
+    fail::cfg("flush_before_cluse_threshold", "return(0)").unwrap();
```
Comment (overvenus) [inline tests/failpoints/cases/test_merge.rs:1886]: typo?

```suggestion
    fail::cfg("flush_before_close_threshold", "return(0)").unwrap();
```

**Review by overvenus — COMMENTED** [review]
Rest LGTM

File: components/raftstore-v2/src/fsm/peer.rs (hunk 1 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/fsm/peer.rs:199]: Using schedule_tick delays the check for some time. I prefer to check immediately after restart.

File: components/raftstore-v2/src/raft/peer.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/raft/peer.rs:234]: Done

**Comment by SpadeA-Tang** [comment]
PTAL again @overvenus Thanks.

**Review by overvenus — APPROVED** [review]

**Pull request merged**